From 7052a19982ac177a740cbbc8d98499854d1cbfaa Mon Sep 17 00:00:00 2001
From: Maciej Fijalkowski <maciej.fijalkowski@intel.com>
Date: Wed, 2 Oct 2024 12:31:19 +0200
Subject: [PATCH 1/4] bpf: add bpf_verifier_env pointer to bpf_prog_aux struct

In order to be able to use the stack within convert_ctx_access callbacks
and avoid touching the its declaration and therefore unnecessary code
churn, let us introduce a pointer to bpf_verifier_env onto bpf_prog_aux.
convert_ctx_access has bpf_prog struct handy so what is here achieved is
the fact that we can use the subprog_info from env to grab the stack
depth from the subprogram that patched insn belongs to and modify/write
it respectively.

Move the num_exentries to existing hole so that we don't increase the
size of modified struct.

Before addition:
/* size: 1088, cachelines: 17, members: 69 */
/* sum members: 1070, holes: 4, sum holes: 18 */
/* paddings: 1, sum paddings: 7 */
/* forced alignments: 2, forced holes: 1, sum forced holes: 4 */

After:
/* size: 1088, cachelines: 17, members: 70 */
/* sum members: 1078, holes: 2, sum holes: 10 */
/* paddings: 1, sum paddings: 7 */
/* forced alignments: 2 */

Scope the pointer only for lifetime of bpf_check().

Signed-off-by: Maciej Fijalkowski <maciej.fijalkowski@intel.com>
---
 include/linux/bpf.h   | 3 ++-
 kernel/bpf/verifier.c | 2 ++
 2 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/include/linux/bpf.h b/include/linux/bpf.h
index 3b94ec161e8c..9ae37bf667a8 100644
--- a/include/linux/bpf.h
+++ b/include/linux/bpf.h
@@ -1480,6 +1480,7 @@ struct bpf_prog_aux {
 	struct bpf_kfunc_desc_tab *kfunc_tab;
 	struct bpf_kfunc_btf_tab *kfunc_btf_tab;
 	u32 size_poke_tab;
+	u32 num_exentries;
 #ifdef CONFIG_FINEIBT
 	struct bpf_ksym ksym_prefix;
 #endif
@@ -1527,7 +1528,7 @@ struct bpf_prog_aux {
 	 */
 	u32 linfo_idx;
 	struct module *mod;
-	u32 num_exentries;
+	struct bpf_verifier_env *env;
 	struct exception_table_entry *extable;
 	union {
 		struct work_struct work;
diff --git a/kernel/bpf/verifier.c b/kernel/bpf/verifier.c
index d8520095ca03..05b7eebecaac 100644
--- a/kernel/bpf/verifier.c
+++ b/kernel/bpf/verifier.c
@@ -21594,6 +21594,7 @@ int bpf_check(struct bpf_prog **prog, union bpf_attr *attr, bpfptr_t uattr, __u3
 	for (i = 0; i < len; i++)
 		env->insn_aux_data[i].orig_idx = i;
 	env->prog = *prog;
+	(*prog)->aux->env = env;
 	env->ops = bpf_verifier_ops[env->prog->type];
 	env->fd_array = make_bpfptr(attr->fd_array, uattr.is_kernel);
 
@@ -21710,6 +21711,7 @@ int bpf_check(struct bpf_prog **prog, union bpf_attr *attr, bpfptr_t uattr, __u3
 		/* program is valid, convert *(u32*)(ctx + off) accesses */
 		ret = convert_ctx_accesses(env);
 
+	(*prog)->aux->env = NULL;
 	if (ret == 0)
 		ret = do_misc_fixups(env);
 
-- 
2.25.1

